rfcs/41-ui-navigation.md

```diff
@@ -0,0 +1,364 @@
+when that `Focusable` is focused and the `Action` request is received,
+the focus changes to the `TreeMenu`
+
+In short, the navigation tree is built on two elements:
+- The `reachable_from` that a `NavMenu` was spawned with
+- The `Focusable` within the hierarchy of a `NavMenu`
+
+This is a tree in the ECS.
+
+In the following screenshot, `Focusable`s are represented as circles, and menus
+as rectangles.
+
+![A screenshot of a RPG-style menu with the navigation tree overlayed](https://user-images.githubusercontent.com/26321040/141671969-ea4a461d-0949-4b98-b060-c492d7b896bd.png)
+
+A `Focusable` used to go from the root menu
+to the menu in which the currently focused entity is is _active_.
+The current focus position is represented as a trail of breadcrumb active
+focusables and the currently focused focusable.
+
+This is reflected into the `NavEvent::FocusChanged` event.
+When going from a focusable to another,
+the navigation system emits a `FocusChanged` event,
+the event contains two lists of entities:
+- The list of entities that after the navigation event are no more focused
+  or active
+- The list of entities that are newly active or focused
+
+For the simplest case of navigating from one focusable to another
+in the same menu, the two list have a single element,
+so it is capital that it is easy to access those elements.
+This is why we use a `NonEmpty`, from the `non_empty_vec` crate.
+A `NonEmpty<T>` has a `first` method that returns a `T`,
+which in our case would be the last and new focused elements.
+I chose this crate over similar ones,
+because the code is dead simple and easy to vet.
+
+#### Going back to a previous menus
+
+`Focusable`s have a `dormant` state that is set when they go from
+active/focused to not active.
+This is a form of memory that allows focus to go back to the last focused
+element within a menu when it is re-visited.
+This is also the mechanism used to let the user decide which focusable to focus
+when none are focused yet.
+
+#### Spawning menus
+
+Menus store a reference to their parent,
+but that parent is not necessarily their hierarchical parent.
+The parent is just a button in another menu.
+It is inconvenient to have to pre-spawn each button to acquire their `Entity` id
+just to be able to spawn the menu you'll get to from that button.
+`ui-navigation` uses a proxy component holding both the parent and the menu state info.
+That proxy component is `MenuSeed`,
+you can initialize it either with the `Name` or `Entity` of the parent focusable of the menu.
+A system will take all `MenuSeed` components and replace them with a `TreeMenu`.
+
+This introduces a single frame lag on menu spawn,
+but it improves ergonomics to users.
+I thought it was a pretty good trade-off,
+since spawning menus is not time critical, unlike input.
+
+
+#### cancel events
+
+When entering a menu, you want a way to leave it and return to the previous menu.
+To do this, we have two options:
```

Comment:
Why do we need both options? What's the difference?

Reply:
A "go back to previous menu" button is a very common pattern, and accommodating it is fairly neat.

It's not too hard for users to implement their own back button, but then it requires them to actual understand the low level navigation API we are exposing here.

Reply:
Seems reasonable :) Explain that context here!